src/policy/fees_input.cpp

```diff
@@ -0,0 +1,222 @@
+        *log << value.write() << std::endl;
+    }
+}
+
+void FeeEstInput::processBlock(unsigned int block_height, const AddTxsFn& add_txs)
+{
+    UniValue json_txs{UniValue::VARR};
+    estimator.processBlock(block_height, [&](const AddTxFn& add_tx) {
+        add_txs([&](const uint256& hash, unsigned int height, CAmount fee, size_t size) {
+            add_tx(hash, height, fee, size);
+            if (log) json_txs.push_back(TxLog(hash, height, fee, size));
+        });
+    });
+    if (log) {
+        UniValue json(UniValue::VOBJ);
+        UniValue json_block(UniValue::VOBJ);
+        json_block.pushKV("height", int64_t(block_height));
+        json.pushKV("block", json_block);
+        json.pushKV("txs", std::move(json_txs));
+        json.pushKV("time", GetTime());
+        *log << json.write() << std::endl;
+    }
+}
+
+void FeeEstInput::removeTx(const uint256& hash, bool in_block)
+{
+    estimator.removeTx(hash, in_block);
+
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        UniValue removeTx(UniValue::VOBJ);
+        removeTx.pushKV("hash", hash.ToString());
+        removeTx.pushKV("inBlock", UniValue(in_block));
+        value.pushKV("removeTx", removeTx);
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+}
+
+void FeeEstInput::flushUnconfirmed()
+{
+    estimator.FlushUnconfirmed();
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        UniValue flush(UniValue::VARR);
+        value.pushKV("flush", flush);
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+}
+
+bool FeeEstInput::writeData(const fs::path& filename)
+{
+    CAutoFile file(fsbridge::fopen(filename, "wb"), SER_DISK, CLIENT_VERSION);
+    if (file.IsNull()) {
+        LogPrintf("%s: Failed to write fee estimates to %s\n", __func__, filename.string());
+        return false;
+    }
+    estimator.Write(file);
+    return true;
+}
+
+bool FeeEstInput::readData(const fs::path& filename)
+{
+    CAutoFile file(fsbridge::fopen(filename, "rb"), SER_DISK, CLIENT_VERSION);
+    if (file.IsNull() || !estimator.Read(file)) {
+        return false;
+    }
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        std::ifstream data(filename.string(), std::ifstream::binary);
+        value.pushKV(
+            "read", HexStr(std::string(std::istreambuf_iterator<char>(data), std::istreambuf_iterator<char>())));
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+    return true;
+}
+
+bool FeeEstInput::writeLog(const std::string& filename)
+{
+    if (log) {
+        UniValue value(UniValue::VOBJ);
+        value.pushKV("stop", Params().NetworkIDString());
+        value.pushKV("time", GetTime());
+        *log << value.write() << std::endl;
+    }
+
+    if (filename.empty()) {
+        log.reset();
+    } else {
+        log = MakeUnique<std::ofstream>(filename, std::ofstream::out | std::ofstream::app);
```

Comment:
Please [use `std::make_unique` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/util/memory.h#L13).